### PR TITLE
ddpm scheduler variance fixes

### PR DESCRIPTION
**Pull request by williamberman**
see PR comments

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 1]
```diff
@@ -214,16 +214,17 @@ def _get_variance(self, t, predicted_variance=None, variance_type=None):
         # and sample from it to get previous sample
         # x_{t-1} ~ N(pred_prev_sample, variance) == add variance to pred_sample
         variance = (1 - alpha_prod_t_prev) / (1 - alpha_prod_t) * current_beta_t
+        variance = torch.clamp(variance, min=1e-20)
```
Comment (williamberman) [inline src/diffusers/schedulers/scheduling_ddpm.py:217]: we always take the log of `variance` so we should clamp it once

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 2]
```diff
@@ -234,7 +235,7 @@ def _get_variance(self, t, predicted_variance=None, variance_type=None):
             return predicted_variance
         elif variance_type == "learned_range":
             min_log = torch.log(variance)
-            max_log = torch.log(self.betas[t])
+            max_log = torch.log(current_beta_t)
```
Comment (williamberman) [inline src/diffusers/schedulers/scheduling_ddpm.py:238]: when skipping timesteps, self.betas[t] is the wrong beta. see line 211 where we re-calculate the beta for the current timestep and previous timestep

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_ddpm.py:217]: Numerical precisions galore :)

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 3]
```diff
@@ -214,16 +214,17 @@ def _get_variance(self, t, predicted_variance=None, variance_type=None):
...
             variance_type = self.config.variance_type
 
         # hacks - were probably added for training stability
         if variance_type == "fixed_small":
-            variance = torch.clamp(variance, min=1e-20)
+            variance = variance
         # for rl-diffuser https://arxiv.org/abs/2205.09991
         elif variance_type == "fixed_small_log":
-            variance = torch.log(torch.clamp(variance, min=1e-20))
+            variance = torch.log(variance, min=1e-20)
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_ddpm.py:227]: Do we still need the `min` here?

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 2 above)
Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_ddpm.py:238]: Brutal.

**Review by sayakpaul — APPROVED** [review]
Good for me to merge given the tests pass.

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 2 above)
Comment (williamberman) [reply src/diffusers/schedulers/scheduling_ddpm.py:238]: @sayakpaul two days of debugging 🙃

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 3 above)
Comment (williamberman) [reply src/diffusers/schedulers/scheduling_ddpm.py:227]: ah thanks!

Comment (williamberman) [reply src/diffusers/schedulers/scheduling_ddpm.py]: https://github.com/huggingface/diffusers/pull/3006/commits/11fbf66e373e425b09649ce14251eee028289201

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 1 above)
Comment (williamberman) [reply src/diffusers/schedulers/scheduling_ddpm.py:219]: https://github.com/huggingface/diffusers/pull/3006/commits/deaf1d614747e2112afab49d5d37bf363c611521

**Review by patrickvonplaten — APPROVED** [review]
Nice fixes

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 4]
```diff
@@ -215,15 +215,18 @@ def _get_variance(self, t, predicted_variance=None, variance_type=None):
...
+        # we always take the log of variance, so clamp it to ensure it's not 0
+        variance = torch.clamp(variance, min=1e-20)
+
         if variance_type is None:
             variance_type = self.config.variance_type
 
         # hacks - were probably added for training stability
         if variance_type == "fixed_small":
-            variance = torch.clamp(variance, min=1e-20)
+            variance = variance
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_ddpm.py:226]:
```suggestion
pass
```

Or remove this `if` condition.
